kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/datatest/names.kt

```diff
@@ -0,0 +1,44 @@
+package io.kotest.core.datatest
+
+import io.kotest.mpp.bestName
+import io.kotest.mpp.reflection
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+
+/**
+ * Each test name must be unique. We can use the toString if we determine the instance is stable.
+ *
+ * An instance is considered stable if it is a data class where each parameter is either a data class itself,
+ * or a primitive type.
+ *
+ * Note: If the user has overridden toString() and the returned value is not stable, tests may not appear.
```

Comment:
Can we somehow notify users about this behavior? For example, if the generated identifier is not unique we can log a warning

Reply:
Excellent idea, or maybe even not support it.

Reply:
As another option, we can concatenate counter to the identifier
Something like `StableIdentifier`, `StableIdentifier (1)`, `StableIdentifier (2)` for repeated values

Reply:
So add the counter even for stable values when repeated.

Reply:
![image](https://user-images.githubusercontent.com/743706/97776835-a594a480-1b39-11eb-90c0-dadda5dec055.png)

![image](https://user-images.githubusercontent.com/743706/97776884-ebea0380-1b39-11eb-99ba-812b6203eaf5.png)

Reply:
If I were a user of such an API, it would be the most intuitive outcome from all the discussed options. 
Will it work the same way for other isolation modes? Do we guarantee the order?  I guess it can be confusing if identifiers are flicky between launches. What will happen in case `Foo(name=ham) (1)` fails and I want to relaunch the failed test only? Will it be named `Foo(name=ham)`? 
What I care about is to identify the place of failure and reproduce it when I have a test report only.

However, I cannot imagine the real-world scenario for this case and consider it as a failover for human mistake. We might want to wait for users' feedback until considering these questions.

Reply:
Good questions. The plugin is not able to re-launch a single test, and probably won't be (as it involves knowing runtime information). The order of execution is not guaranteed if using different isolation modes, but the ordering of names is as all the tests are guaranteed at once.